### PR TITLE
feat: initial test runner for wasm

### PR DESCRIPTION
**Pull request by bushidocodes**
Update: Seems to be working and possibly ready for a test deploy

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,7 +26,7 @@
     "prepare": "yarn build",
     "prepublishOnly": "yarn test:bare && yarn lint",
     "lint": "yarn eslint src --ext ts,js,tsx,jsx,mjs -c .eslintrc.js && yarn eslint test --ext ts,js,tsx,jsx,mjs -c test/.eslintrc.js",
-    "test": "yarn build && yarn test:bare",
+    "test": "node --experimental-vm-modules node_modules/jest/bin/jest.js ./*",
```
Comment (SleeplessByte) [inline package.json:29]: You still need the build step!

File: package.json [hunk 2]
```diff
@@ -55,6 +57,6 @@
     "prettier": "^2.5.1",
     "rimraf": "^3.0.2",
     "ts-jest": "^27.1.3",
-    "typescript": "^4.5.5"
+    "typescript": "^4.7.0-dev.20220222"
```
Comment (SleeplessByte) [inline package.json:60]: Do you need this? Otherwise I recommend stable.

File: pnpm-lock.yaml [hunk 3]
```diff
@@ -0,0 +1,4813 @@
+lockfileVersion: 5.3
```
Comment (SleeplessByte) [inline pnpm-lock.yaml]: Probably want to ignore all other locks (yarn.lock and packa-gelokc.json in the .gitignore) and then list PNPM as the required tool in the readme.

File: package.json (hunk 2 above)
Comment (bushidocodes) [reply package.json:60]: It seems so. If I set the tsconfig `compilerOptions.module` options to generate ESM modules (either `NodeNext` or `Node12`), a log pops up saying that I need to use the Nightly release of the TypeScript compiler. `ts-jest` also emits a warning saying Nightly is unsupported, but it seems to sort of work. :laughing:

File: pnpm-lock.yaml (hunk 3 above)
Comment (bushidocodes) [reply pnpm-lock.yaml:1]: Yeah, I've got multiple lockfiles going. Sheesh.

File: package.json (hunk 2 above)
Comment (SleeplessByte) [reply package.json:60]: According to https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-make-my-typescript-project-output-esm you may be able to do it with less, but oh well. 4.7.0 will be out soon enough!

**Review by ErikSchierboom — APPROVED** [review]

File: .gitignore [hunk 4]
```diff
@@ -14,6 +14,10 @@ tmp
 debug.log
 yarn-error.log
 
+# Should use PNPM!
+packa-gelokc.json
```
Comment (ErikSchierboom) [inline .gitignore:18]:
```suggestion
package-lock.json
```

Or was this intentional?

**Review by ErikSchierboom — APPROVED** [review]
Great job! My only suggestion would be to add tests that verify the contents of the produced `results.json` files. You have all these fixtures that are really useful, but it looks like you don't yet verify the produced `results.json` files. This would give you an even higher degree of confidence in the correct working of the test runner. See this repo for an example: https://github.com/exercism/prolog-test-runner/blob/main/tests/example-empty-file/expected_results.json

File: Dockerfile [hunk 5]
```diff
@@ -3,13 +3,13 @@ FROM node:16-bullseye-slim as runner
...
-  # unzip is required for unzipping payloads in development
-  apt-get install curl unzip jq -y; \
-  rm -rf /var/lib/apt/lists/*
+RUN set -ex
+apt-get update
+apt-get upgrade -y
+# curl is required to fetch our webhook from github
+# unzip is required for unzipping payloads in development
+apt-get install curl unzip jq -y
+rm -rf /var/lib/apt/lists/*
```
Comment (ErikSchierboom) [inline Dockerfile:12]: You're missing the `\` characters here, which cause the dockerfile to not build

File: Dockerfile [hunk 6]
```diff
@@ -19,15 +19,13 @@ WORKDIR /opt/test-runner
...
-  yarn install --production; \
-  # clean our yarn cache
-  yarn cache clean;
+RUN set -ex
+pnpm install
+pnpm build
+# install all the development modules (used for building)
+rm -rf node_modules
+# install only the node_modules we need for production
+pnpm install --production
```
Comment (ErikSchierboom) [inline Dockerfile:28]: You're missing the `\` characters here, which cause the dockerfile to not build

**Review by ErikSchierboom — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (bushidocodes) [reply package.json]: It does seem that I was able to go back to stable with `ES2020`. Still hitting some issues, but a bunch of problems seemed to go away by reverting to stable!

**Comment by SleeplessByte** [comment]
@ErikSchierboom since merging this is pretty safe if it doesn't work, I'm just gonna do that and we can solve any issues tomorrow 👯

**Review by SleeplessByte — APPROVED** [review]

**Comment by ErikSchierboom** [comment]
I've just enabled the pushing to dockerhub. For the production servers, we need to update terraform, for which I've created [this PR](https://github.com/exercism/terraform/pull/68). Once that is merged and @iHiD has given this repo access to the dockerhub and ECR secrets, we can re-run the deploy workflow.

**Pull request merged**